### PR TITLE
TRT-2633: Link symptoms to triage records with summaries and filtering

### PR DESCRIPTION
**Pull request by smg247**
Surfaces job run symptoms on triage detail pages by syncing them into a
  triage_symptoms junction table during regression cache loading, then
  exposing per-symptom summaries (with regression IDs and job run counts)
  via the expand=symptoms query parameter. Adds frontend symptom chips on
  regression rows and a filterable symptom summary panel. Includes unit
  and e2e tests for symptom syncing, API expansion, and cascade deletion.

Triage Details page with Symptoms:
Screenshot 2026-05-05 at 7 31 37 PM

Filtered:
Screenshot 2026-05-05 at 7 31 51 PM

Audit logging was left out of this effort. In the future, symptoms might be manually added to the triage. At that time, we would want to consider how to handle the audit logging (likely similar to what we do with regressions).

## Summary by CodeRabbit

* **New Features**
  * Symptoms are now tracked and persisted on triage records; API returns per-symptom summaries with regression counts, job-run counts and percentages. Frontend shows a "Symptoms" section, per-regression symptom chips, and symptom-based filtering.

* **Chores**
  * DB migrations and schema updates to store symptom associations and enable cascade deletes; background sync added to compute and upsert symptom counts idempotently.

* **Tests**
  * New unit and e2e tests for symptom propagation, syncing/idempotency, API expansion, UI filtering, and cascade-delete behavior.

### CODE REVIEW

File: pkg/api/componentreadiness/regressiontracker.go [hunk 1]
```diff
@@ -109,6 +112,10 @@ func (prs *PostgresRegressionStore) MergeJobRuns(regressionID uint, jobRuns []mo
 		jobRuns[i].RegressionID = regressionID
 		res := prs.dbc.DB.
 			Where("regression_id = ? AND prow_job_run_id = ?", regressionID, jobRuns[i].ProwJobRunID).
+			Assign(models.RegressionJobRun{
+				JobLabels:   jobRuns[i].JobLabels,
+				JobSymptoms: jobRuns[i].JobSymptoms,
+			}).
 			FirstOrCreate(&jobRuns[i])
```
[the remark this answers is not in the document]
Comment (smg247) [reply pkg/api/componentreadiness/regressiontracker.go:119]: Those other fields aren't something that will change in subsequent loads, so that is wasteful.

File: pkg/api/componentreadiness/regressiontracker.go [hunk 2]
```diff
@@ -118,6 +125,64 @@ func (prs *PostgresRegressionStore) MergeJobRuns(regressionID uint, jobRuns []mo
...
+				result := prs.dbc.DB.Where(ts).FirstOrCreate(&ts)
+				if result.Error != nil {
+					return fmt.Errorf("error syncing symptom %s to triage %d regression %d: %w",
+						symptomID, triage.ID, reg.ID, result.Error)
+				}
+				if err := prs.dbc.DB.Model(&ts).Update("job_run_count", count).Error; err != nil {
+					return fmt.Errorf("error updating symptom job run count: %w", err)
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (smg247) [reply pkg/api/componentreadiness/regressiontracker.go:177]: This would only happen if a symptom itself was deleted. I have added a cascade on delete to account for this.

[the remark this answers is not in the document]
Comment (smg247) [reply pkg/api/componentreadiness/regressiontracker.go:177]: Sure, but symptoms don't disappear from a job run unless they are deleted

**Review by stbenjam — CHANGES_REQUESTED** [review]
## Review Panel Verdict

**Disposition**: REQUEST_CHANGES — Two data integrity gaps around stale/orphaned `triage_symptoms` rows require fixes before merge.

---

### Specialist Findings

**Architecture Reviewer**:
- SUGGESTION — `SyncTriageSymptoms` uses `FirstOrCreate` + separate `Update` instead of the `Assign` + `FirstOrCreate` pattern already used by `MergeJobRuns` in the same file, doubling DB round-trips per symptom (`regressiontracker.go:165-180`).
- SUGGESTION — `Preload("JobRuns")` for all active regressions could load tens of thousands of rows into memory at once (`regressiontracker.go:142-146`). Consider batching or a SQL-based aggregation.
- SUGGESTION — `GetTriageSymptomSummaries` makes 3 DB queries where 2 would suffice — query #3 re-fetches rows already partially aggregated in query #1 (`triage.go:878-941`).
- NOTE — No FK constraint from `triage_symptoms.regression_id` to `test_regressions.id`, risking orphaned rows on regression deletion.
- NOTE — `regressionSymptomMap` is referenced before its declaration in `TriagedRegressionTestList.js` (works at runtime due to closure semantics but is fragile).
- NOTE — Removal of `backfillClosedRegressionViews` is unrelated cleanup bundled into this PR.
- NOTE — Overall architecture is sound, follows existing patterns well. Error handling is correct throughout.

**Security & Supply Chain Reviewer**:
- SUGGESTION — `TriageSymptomSummary` embeds the full `jobrunscan.Symptom` model, exposing internal matching rules (`MatchString`, `FilePattern`, `MatcherType`) in API responses. Consider a slimmed-down DTO with just `ID` and `Summary` (`triage.go:868`).
- NOTE — `expand` parameter silently ignores unknown values — safe but could mask typos.
- NOTE — Error messages include internal details, consistent with existing codebase patterns.
- NOTE — No dependency changes, no new external calls, no injection vectors, no XSS concerns. Supply chain is clean.

**UX & API Reviewer**:
- SUGGESTION — No validation of invalid `expand` values; `?expand=symtpoms` (typo) silently returns no symptom data (`server.go:1680-1684`).
- SUGGESTION — Stale `triage_symptoms` rows are never cleaned up, leading to inaccurate data over time when symptoms disappear from job runs (`regressiontracker.go:131-184`).
- SUGGESTION — `TriageSymptoms` field on the `Triage` model has `json:"triage_symptoms,omitempty"` — raw junction rows could leak into API responses if someone adds preloading. Consider `json:"-"` (`models/triage.go:52`).
- SUGGESTION — Tooltip on the "Symptoms" heading could be more discoverable with an info icon (`TriageSymptoms.js:26-33`).
- NOTE — Symptom chip truncation at 12 characters is arbitrary; CSS `text-overflow` might adapt better.
- NOTE — Good accessibility practices with `aria-label` and `aria-pressed`. Deterministic `symptomColor` is a good design choice.

**Codebase Consistency Reviewer**:
- SUGGESTION — `seedSymptom` and `cleanupTriageSymptoms` test helpers are duplicated across two e2e packages with inconsistent implementations (one uses `FirstOrCreate`, the other uses `Create`).
- SUGGESTION — Duplicate imports from `'./CompReadyUtils'` in `TriagedRegressionTestList.js` (lines 5 and 8) — should be combined into one.
- SUGGESTION — `SyncTriageSymptoms` uses two-query upsert instead of the `Assign` pattern established by `MergeJobRuns` in the same file.
- SUGGESTION — `RegressedTests` JSON tag missing `omitempty`, inconsistent with `SymptomSummaries` which has it (`server.go:1672`).
- NOTE — `sort.Slice` used instead of `slices.SortFunc` despite `slices` already being imported (`triage.go:938`).
- NOTE — Overall strong consistency with existing patterns in GORM models, error wrapping, BigQuery queries, logging, and frontend conventions.

**QA Engineer**:
- BLOCKING — No unit test for `GetTriageSymptomSummaries`, a non-trivial exported function with aggregation logic, silent-skip behavior, and percentage computation (`triage.go:878`).
- BLOCKING — `SyncTriageSymptoms` never removes stale junction rows. No test covers the scenario where a symptom disappears from all job runs after a sync (`regressiontracker.go:128-184`).
- SUGGESTION — No test for `expand` parameter with unknown/invalid values.
- SUGGESTION — No test for `symptomColor` utility function (`CompReadyUtils.js:914-920`).
- SUGGESTION — No test for duplicate symptom IDs within a single job run's `JobSymptoms` array.
- NOTE — `LinearProgress` percentage could theoretically exceed 100% if data is inconsistent. Consider clamping.
- NOTE — No frontend tests for `TriageSymptoms` component, consistent with existing codebase (no existing tests for these components).

**Devil's Advocate**:
- BLOCKING — No FK from `triage_symptoms.regression_id` to `test_regressions.id`. When a regression is deleted, orphaned `triage_symptoms` rows persist and inflate symptom counts. When a regression is unlinked from a triage, same problem (`models/triage.go:271-276`).
- BLOCKING — `SyncTriageSymptoms` never deletes stale junction rows. If BigQuery returns different symptoms for a job run on a subsequent query (e.g., symptom definition deleted, BQ data corrected), old rows persist with stale counts (`regressiontracker.go:131-184`).
- SUGGESTION — N+1 DB queries: `200 regressions * 3 triages * 5 symptoms * 2 ops = 6000` round-trips per loader run. Use bulk `ON CONFLICT ... DO UPDATE`.
- SUGGESTION — `Preload("JobRuns")` loads all job runs into memory. Consider SQL-based aggregation to avoid the memory spike.
- SUGGESTION — `backfillClosedRegressionViews` removal is unrelated; confirm it has run in all environments.
- NOTE — `SyncTriageSymptoms` includes regressions from errored releases — intentional but symptom counts may be understated.

**Technical Writer**:
- SUGGESTION — Plan doc states symptoms are "also computed when `regressions` is requested" but the implementation handles them independently; frontend explicitly requests both (`docs/plans/trt-2633-triage-symptoms.md:412-419`).
- SUGGESTION — Plan specifies Regressions column should display "2 / 5" (fraction format) but implementation only shows the count.
- SUGGESTION — Plan specifies symptom rows should link to Job Artifact Query page, but implementation renders plain chips with no links.
- NOTE — New public Go functions have appropriate godoc comments. No existing external documentation requires updating.

---

### Panel Synthesis

Three independently-arriving specialists (Architecture, UX/API, Devil's Advocate, QA) all converge on the same two data integrity issues:

1. **`SyncTriageSymptoms` is additive-only and never cleans up stale junction rows.** The design doc claims symptoms "only accumulate," but `MergeJobRuns` uses `Assign` which *replaces* `JobSymptoms` on each sync. If BigQuery data changes, symptom counts become permanently inflated.

2. **No FK constraint from `triage_symptoms.regression_id` to `test_regressions.id`.** When a regression is deleted, orphaned rows persist and inflate symptom counts.

---

### Required Actions Before Merge

1. **Add cleanup of stale `triage_symptoms` rows** in `SyncTriageSymptoms` (`pkg/api/componentreadiness/regressiontracker.go:131-184`). After computing `symptomCounts` for each `(triage, regression)` pair, delete any existing `triage_symptoms` rows whose `symptom_id` is not in the current set.

2. **Add FK constraint from `triage_symptoms.regression_id` to `test_regressions.id` with `ON DELETE CASCADE`** (`pkg/db/db.go`). Follow the existing `ensureTriageSymptomCascade` pattern.

---

### Optional Follow-ups

- Consolidate `FirstOrCreate` + `Update` into `Assign` + `FirstOrCreate` (or batch `ON CONFLICT ... DO UPDATE`) to reduce DB round-trips
- Add unit tests for `GetTriageSymptomSummaries` covering edge cases: zero regressions, stale symptom IDs, percentage rounding, sort order
- Slim down `TriageSymptomSummary` to expose only `ID` and `Summary` from the `Symptom` model rather than the full struct with matching rules
- Add `json:"-"` to the `TriageSymptoms` field on the `Triage` model to prevent raw junction rows from leaking into API responses
- Deduplicate `seedSymptom`/`cleanupTriageSymptoms` test helpers across e2e packages
- Update plan document to reflect implementation deviations (independent expand handling, count-only display, deferred JAQ links)

---
*Review generated by multi-specialist panel (Architecture, Security & Supply Chain, UX/API, Codebase Consistency, QA, Devil's Advocate, Technical Writer + Panel Arbiter)*

**Comment by stbenjam** [comment]
From the panel review, the required requests seem to make sense to me, and the unit test ask seems reasonable if its testable.

@coderabbitai What do you think about the panel review above? ^^

**Comment by stbenjam** [comment]
@coderabbitai

> Required Action 1 (stale triage_symptoms rows): This is the more impactful of the two. SyncTriageSymptoms is purely additive — if BigQuery data changes (a symptom definition is removed, or a job run is re-processed with different symptoms), old rows persist with stale counts. MergeJobRuns already handles this correctly for JobSymptoms by using Assign, which replaces the field on each sync, so there's precedent in this file for the expected behavior.

BigQuery doesn't really ever change, though.

**Comment by smg247** [comment]
This was designed with the assumption that job runs would not stop showing symptoms unless the symptom itself was deleted. I can't think of any reason why we would want to account for that possibility, and I believe it will make the logic more complex to do so.

I will add the FK constraint from `triage_symptoms.regression_id` to `test_regressions.id` as that is certainly desired.

**Comment by smg247** [comment]
Waiting on https://github.com/openshift/sippy/pull/3481 merge to add additional unit testing

**Comment by dgoodwin** [comment]
This looks great, ty!
